Frontend redesign and codebase cleanup

j-james opened the pull request:
Closes #47.

Running todo list:
- [x] Make light header background color opaque
- [x] Scrape package descriptions
- [x] Format descriptions to not over/underflow boxes
- [x] Fix light/dark theme switcher
- [x] Fix strange theme switcher button styling
- [x] Space blocks in the footer evenly
- [x] Find a place for packages.xml
- [x] Overhaul remaining pages (about, logs, etc)
- [x] Clean up codebase
- [x] Scrape package licenses
- [x] Scrape package last-updated times
- [x] Ensure all tags on the homepage have some packages 
- [x] Overhaul `/pkg/` pages
- [x] Remove version info from package cards
- [x] Add a link to the package's repository to package cards

Future todo list:
- [ ] Add last-updated times to package cards
- [ ] Fix package documentation pages
- [ ] Scrape package versions efficiently, somehow
- [ ] Fetch GitHub stats on every search (for searches returning > x results?)

Rendered: [1](https://files.catbox.moe/socejz.png) [2](https://files.catbox.moe/pvv1al.png) [3](https://files.catbox.moe/ncfjye.png) [4](https://files.catbox.moe/cgkiex.png) [5](https://files.catbox.moe/qfzllx.png) [6](https://files.catbox.moe/cb877p.png)

georgelemon commented:
For about page I will create 2 or 3 png artworks to showcase Nimble & Nim. This week I hope <3

georgelemon commented:
Forgot to mention. for more SVG icons you have [Feather](https://feathericons.com/) and [Octicons](https://primer.style/octicons/)

j-james commented:
@FedericoCeratto Mind taking a look at this when you have the chance?

FedericoCeratto commented on package_directory.nim:
```diff
@@ -611,32 +614,23 @@ proc fetch_github_versions(pkg: Pkg, owner_repo_name: string) {.async.} =
+  let q = github_repository_search_tpl % [date, "20", "updated", "1"]
+  log_info "Searching GH repos: '$#'" % q
+  let query_res = await getGHJson(q)
+  # NOTE: getGHJson should have descriptions in it???
```
uh?

j-james replied:
What's weird here, the comment? I think it might have been from my notes while reading though the codebase - it doesn't look it's present in the final diff.

FedericoCeratto reviewed (changes requested):
I left some initial comments. Also, for a website showing a directory of software packages the complex artworks with statues, flowers, zebras etc are a bit unnecessary and probably overwhelming and confusing to some users. Can we please remove them?

j-james commented on friendly_timeinterval.nim:
```diff
@@ -1,10 +1,9 @@
-import strutils
+import std/[times, strutils]
 
+## Convert TimeInterval to human-friendly description
```
Whoops - reverted, along with some other comment changes.

j-james commented on github.nim:
```diff
@@ -3,60 +3,179 @@
 # GitHub interface
 #
 
```
Alright, I rewrote the commit history to be more granular (from the code side. for the html templates, i merged the in-development fixes commits together.)

j-james commented:
CC. @georgelemon for the artwork request. I personally like it a lot, but it's true that it doesn't have much to do with Nim. It might be fun if we did something involving pixie's image generation?

j-james commented:
@FedericoCeratto The commit history has been rewritten and I fixed a few more bugs. Does the rest of the code look good?

j-james commented:
(bump: @FedericoCeratto)

FedericoCeratto commented:
@j-james The code changes are ok, thank you! I was waiting for updates on the artwork.
For context, other forges or package indexes including GitHub, Sourcehut, Gitea, packages.debian.org, PyPI, Guix, repology.org do not have artwork beyond a logo/favicon.

moigagoo commented:
> CC. @georgelemon for the artwork request. I personally like it a lot, but it's true that it doesn't have much to do with Nim. It might be fun if we did something involving pixie's image generation?

This is probably something that can be safely done later.

moigagoo commented:
> @j-james The code changes are ok, thank you! I was waiting for updates on the artwork. For context, other forges or package indexes including GitHub, Sourcehut, Gitea, packages.debian.org, PyPI, Guix, repology.org do not have artwork beyond a logo/favicon.

Maybe we could give it a shot with the bold crazy artwork? If it works poorly, we can always reduce its amount.

moigagoo commented:
> Thank you very much for the PR. I left some initial comments. Also, for a website showing a directory of software packages the complex artworks with statues, flowers, zebras etc are a bit unnecessary and probably overwhelming and confusing to some users. Can we please remove them?

I'll repeat my opinion from another place.

Although abundance of bright artwork may be confusing to some, it also may be entertaining and attractive to others. We won't know until we try it.

I think it's better to release and then update than keep it unreleased forever.

j-james commented:
(ping: @FedericoCeratto)

FedericoCeratto commented:
@j-james If you want to have small bugfixes merged sooner the best way would be to open a small PRs with only tiny, targeted commits, without UI redesign and without medium/large code refactors.
The big UI redesign can be then finished and merged later on.

j-james commented:
@FedericoCeratto The big UI redesign is finished from my perspective. The code refactoring changes _are_ in their own commits, if you don't want to take them, but IMO they clean up the codebase significantly and are less scary than the diff may seem (a lot of moving functions to smaller files and merging duplicated functionality).

j-james commented:
(@FedericoCeratto bump?)

j-james commented:
@FedericoCeratto 
This PR has been tested and ready to merge for **four months**. I do not understand what you are expecting to happen.
I believe we addressed all of your comments.

It is not a PR about small bugfixes. It is a stylistic redesign closing #47, which you appeared to be interested in. As for the large codebase refactoring - again, I'll implore you to look at the commits - the refactoring is straightforward and mostly a matter of removing duplicated functionality and putting functions in more appropriate places. The github.nim file, for example, hasn't been touched in *six years* and contained a single function not used anywhere in the codebase.
